pkg/trace/writer/newwriter/sender.go

```diff
@@ -0,0 +1,382 @@
+
+	wg     sync.WaitGroup // waits for all uploads
+	climit chan struct{}  // acts as a semaphore for limiting concurrent connections
+
+	mu        sync.Mutex  // guards below fields
+	list      *list.List  // send queue
+	size      int         // size of send queue
+	scheduled bool        // reports if a flush is scheduled
+	attempt   int         // retry attempt coming up next
+	timer     *time.Timer // timed flush (triggered by backoff)
+}
+
+// senderConfig specifies the configuration for the sender.
+type senderConfig struct {
+	// client specifies the HTTP client to use when sending requests.
+	client *http.Client
+	// url specifies the URL to send requests too.
+	url *url.URL
+	// apiKey specifies the Datadog API key to use.
+	apiKey string
+	// maxConns specifies the maximum number of allowed concurrent ougoing
+	// connections.
+	maxConns int
+	// recorder specifies the eventRecorder to use when reporting events occurring
+	// in the sender.
+	recorder eventRecorder
+}
+
+// newSender creates a new sender, which uses the given HTTP client, URL and API key
+// to make requests using incoming payloads. The limit of outgoing concurrent connections
+// is defined by climit.
+func newSender(cfg *senderConfig) *sender {
+	return &sender{
+		cfg:    cfg,
+		climit: make(chan struct{}, cfg.maxConns),
+		list:   list.New(),
+	}
+}
+
+// Push pushes p onto the sender, to be written to the destination.
+func (q *sender) Push(p *payload) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// enqueue the payload
+	q.enqueueLocked(p)
+
+	if !q.scheduled {
+		// no flush is scheduled; start one
+		q.scheduled = true
+		q.wg.Add(1)
+		go q.flush()
+	}
+}
+
+// flush drains and sends the entire queue. If anything comes in while flushing
+// or if some of the payloads fail to send as retriable, further follow up flushes
+// are scheduled.
+func (q *sender) flush() {
+	startTime := time.Now()
+
+	// we drain the queue, which is a blocking operation, meaning that
+	// new payloads are stopped from coming in while this happens; but
+	// it is a fast.
+	payloads := q.drainQueue()
+
+	// we send the payloads we've retrieved; while we do this, more payloads
+	// can join the list.
+	done, retries := q.sendPayloads(payloads)
+
+	if done > 0 {
+		q.recordEvent(eventTypeFlushed, &eventData{
+			count:    int(done),
+			duration: time.Since(startTime),
+		})
+	}
+
+	// we reassess the state of the list and check if further flushes need
+	// to be scheduled
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	defer q.wg.Done()
+
+	if q.list.Len() == 0 {
+		// the list is empty; no further flushing needs to be triggered
+		q.attempt = 0
+		q.scheduled = false
+		return
+	}
+	// the list is not empty
+	q.scheduled = true
+	if retries > 0 {
+		// some sends failed as retriable; we need to back off a bit
+		q.attempt++
+		q.wg.Add(1)
+		q.timer = time.AfterFunc(backoffDuration(q.attempt), q.flush)
+		return
+	}
+	// all items in the list are new; flush immediately
+	q.wg.Add(1)
+	go q.flush()
```

Comment:
Nit: Given this structuring of `flush`, we'll only ever have 1 routine actively on it. Sounds like we might be able to achieve a cleaner implementation if this routine is always the same and simply does something like:

```
for _ := range q.flushReq {
  backoffDuration := flush()
  time.Sleep(backoffDuration)
}
```

With `q.flushReq = make(chan bool, 1)`.

And wherever you are doing `go q.flush()` we could just call:

```
func (q *sender) requestFlush() {
  select {
    case q.flushReq <- true:
       // cool beans
    default:
       // req already there so no point in adding another
}
```

Reply:
The current implementation was a bit inspired from other projects (e.g. Opencensus) and partly motivated by the fact that we have way too many channels and goroutines in the trace-agent. My aim was to avoid using them, unless needed. In a way, it does make sense to have the goroutine you suggest, but from what I've seen in some profiles and given that we already know that the runtime scheduler is one of the next bottlenecks to address in the agent, I didn't want to put more pressure on it.

To clarify, I think that having a goroutine waiting for work could determine the scheduler to "occasionally visit it" to check if it's ready for work. And many times it won't be, so it will be wasted time. This type of "visiting" for nothing already happens quite often. 

At the same time, it is of course debatable that the "always freshly" created goroutines have a similar impact due to calls like `runtime.morestack` etc (which may be optimized in this case because the call is a method; I'm not sure). At the very least, in this scenario, we know for certain that the goroutine which was started has work to do and isn't just idle waiting around.

I would leave it as it is.